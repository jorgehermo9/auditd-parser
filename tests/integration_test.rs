use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
