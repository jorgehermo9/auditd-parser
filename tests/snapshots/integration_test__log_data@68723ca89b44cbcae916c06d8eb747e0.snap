---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/linux-audit-parser-rs/line-user-logout.log:1"
expression: result
info: "type=USER_LOGOUT msg=audit(1725003193.331:117292): pid=1712 uid=0 auid=1000 ses=764 msg='op=login id=1000 exe=\"/usr/sbin/sshd\" hostname=other-host.invalid addr=10.10.10.10 terminal=ssh res=success'\u001dUID=\"root\" AUID=\"ec2-user\" ID=\"ec2-user\""
---
{
  "record_type": "USER_LOGOUT",
  "timestamp": 1725003193331,
  "id": 117292,
  "fields": {
    "auid": 1000,
    "msg": {
      "addr": "10.10.10.10",
      "exe": "/usr/sbin/sshd",
      "hostname": "other-host.invalid",
      "id": 1000,
      "op": "login",
      "res": "success",
      "terminal": "ssh"
    },
    "pid": "1712",
    "ses": "764",
    "uid": "root"
  },
  "enrichment": {
    "AUID": "ec2-user",
    "ID": "ec2-user",
    "UID": "root"
  }
}
